tiledb/api/src/array/mod.rs

```diff
@@ -367,6 +371,291 @@ impl Array {
+                        ctx,
+                        c_array,
+                        c_name.as_ptr(),
+                        start.as_mut_ptr() as *mut std::ffi::c_void,
+                        end.as_mut_ptr() as *mut std::ffi::c_void,
+                        &mut c_is_empty,
+                    )
+                })?;
+
+                if c_is_empty == 1 {
+                    unreachable!("Non-empty domain was non-empty for size check but empty when retrieving data: dimension = {:?}, start_size = {}, end_size = {}",
+                            c_name, start_size, end_size)
+                } else {
+                    Ok(Some(VarValueRange::from((start, end))))
+                }
+            }
+        }
+    }
+
+    fn dimension_nonempty_domain_impl(
+        &self,
+        dimension_key: LookupKey,
+        datatype: Datatype,
+        cell_val_num: CellValNum,
+    ) -> TileDBResult<Option<TypedRange>> {
+        match cell_val_num {
+            CellValNum::Fixed(nz) => {
+                fn_typed!(datatype, LT, {
+                    type DT = <LT as LogicalType>::PhysicalType;
+                    Ok(self
+                        .dimension_nonempty_domain_impl_fixed::<DT>(
+                            dimension_key,
+                            nz,
+                        )?
+                        .map(|range| TypedRange { datatype, range }))
+                })
+            }
+            CellValNum::Var => {
+                fn_typed!(datatype, LT, {
+                    type DT = <LT as LogicalType>::PhysicalType;
+                    let var_range = self
+                        .dimension_nonempty_domain_impl_var::<DT>(
+                            dimension_key,
+                            datatype,
+                        )?;
+                    Ok(var_range.map(|var_range| TypedRange {
+                        datatype,
+                        range: Range::Var(var_range),
+                    }))
+                })
+            }
+        }
+    }
+
+    /// Returns the non-empty domain of a dimension from this array, if any.
+    /// The "non-empty domain" is the set of coordinates
+    /// which have had data written into them.
+    ///
+    /// This is the union of the non-empty domains of all array fragments.
+    pub fn dimension_nonempty_domain(
+        &self,
+        dimension_key: impl Into<LookupKey>,
+    ) -> TileDBResult<Option<TypedRange>> {
+        let key = dimension_key.into();
+        let (datatype, cell_val_num) = {
+            let dim = self.schema()?.domain()?.dimension(key.clone())?;
+            (dim.datatype()?, dim.cell_val_num()?)
+        };
+        self.dimension_nonempty_domain_impl(key, datatype, cell_val_num)
+    }
+
+    /// Returns the non-empty domain of all dimensions from this array, if any.
+    /// The "non-empty domain" is the set of coordinates across all dimensions
+    /// which have had data written into them.
```

Comment:
This comment and the one above seem a bit oddly worded to me. I wouldn't call it the "set" of coordinates, though two does technically count there.

How about something along the lines of "The domain of a dimension is set at schema creation time. The non-empty domain is the minimum and maximum coordinate value that has been written to disk so far."

Reply:
I took this suggestion and added a bit more to it.

This is why I think it's important we do more of the docs as we go - they will each get more attention than they would with one big PR before publishing time :)